### PR TITLE
Test properties with empty tags

### PR DESCRIPTION
**Pull request by Arthasasyan**

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.axibase.tsd.api.method.property.PropertyTest.assertPropertyExisting;
+import static com.axibase.tsd.api.util.TestUtil.getCurrentDate;
+import static org.testng.AssertJUnit.assertTrue;
+
+public class EmptyTagTest extends PropertyMethod {
+
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Too many blank space

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Issue("6234")
+    @Test
+    public void emptyTagFailTest() {
+        Property property = new Property("property-command-emptytag-test-t1", "property-command-emptytag-test-e1");
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Use Mocks class to generate unique property type and entity

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        Map<String, String> tagMap = new HashMap<>();
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Use ImmutableMap for tags

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        tagMap.put("t1", "\r");
+        property.setTags(tagMap);
+        property.setDate(getCurrentDate());
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Provide only the fields that influence on result. If field is optional, don't include it

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        PlainCommand command = new PropertyCommand(property);
+        CommandSendingResult result = CommandMethod.send(command);
+        tagMap.remove("t1");
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: ?

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        property.setTags(tagMap);
+        assertTrue("The only command had to fail. Actual response: " + result.toString(), result.getFail() == 1 && result.getTotal() == 1);
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: "Single assertion per test" principle doesn't work here. Test different counters in separate statements.

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 1 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Removed

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+import static com.axibase.tsd.api.method.property.PropertyTest.assertPropertyExisting;
+import static com.axibase.tsd.api.util.TestUtil.getCurrentDate;
+import static org.testng.AssertJUnit.assertTrue;
+
+public class EmptyTagTest extends PropertyMethod {
+
+
+    @Issue("6234")
+    @Test
```
[the remark this answers is not in the document]
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Added description annotations for methods:

```java
@Description("Tests that if the only tag in insertion command is empty, it fails.")
    public void emptyTagFailTest()
```

```java
@Description("Tests that if one tag is empty and one is not, property is inserted.")
    public void emptyAndNonEmptyTagTest()
```

```java
@Description("Tests that if key is not empty and the only tg is empty, property is not inserted.")
    public void keyAndEmptyTagTest()
```

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 2 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Done

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 5 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Tag removal was an artefact from previous versions. Deleted.

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 4 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Removed

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 3 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Done

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 6 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Divided into different assertions

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+public class EmptyTagTest extends PropertyMethod {
+
+    @Issue("6234")
+    @Description("Tests that if the only tag in insertion command is empty, it fails.")
+    @Test
+    public void emptyTagFailTest() {
+        String propertyType = Mocks.propertyType();
+        String entityName = Mocks.entity();
+        Property property = new Property(propertyType, entityName);
+        Map<String, String> tagMap = ImmutableMap.<String, String>builder()
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: ImmutableMap.of is clearer

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                .put("t1", "\r")
+                .build();
+        property.setTags(tagMap);
+        PlainCommand command = new PropertyCommand(property);
+        CommandSendingResult result = CommandMethod.send(command);
+        assertTrue("The only command had to fail. Actual response: " + result.toString(), result.getFail() == 1);
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: assertEquals

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Map;
+
+import static com.axibase.tsd.api.method.property.PropertyTest.assertPropertyExisting;
+import static com.axibase.tsd.api.util.TestUtil.getCurrentDate;
+import static org.testng.AssertJUnit.assertTrue;
+
+public class EmptyTagTest extends PropertyMethod {
+
+    @Issue("6234")
+    @Description("Tests that if the only tag in insertion command is empty, it fails.")
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Test description doesn't match test content: empty tag value in description, carriage return in test case

Comment (raipc) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Create a data provider with different values evaluated to empty after trim, e.g. "", "\r", "\n", "\r\n", " ", " \n "

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 9 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Changed for command tests, can not change for tcp tests

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 10 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Data provider is created. Can description be left like this?

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 11]
```diff
@@ -10,81 +10,93 @@
...
-        Map<String, String> tagMap = ImmutableMap.<String, String>builder()
-                .put("t1", "\r")
-                .build();
+        Map<String, String> tagMap = ImmutableMap.of("t1", emptyValue);
         property.setTags(tagMap);
         PlainCommand command = new PropertyCommand(property);
         CommandSendingResult result = CommandMethod.send(command);
-        assertTrue("The only command had to fail. Actual response: " + result.toString(), result.getFail() == 1);
-        assertTrue("Total commands count is not 1: " + result.toString(), result.getTotal() == 1);
+        assertEquals("The only command had to fail. Actual response: " + result.toString(), result.getFail(), Integer.valueOf(1));
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Unnecessary boxing

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    public void emptyAndNonEmptyTagTest(String emptyValue) {
+        String propertyType = Mocks.propertyType();
+        String entityName = Mocks.entity();
+        Property property = new Property(propertyType, entityName);
+        Map<String, String> tagMap = ImmutableMap.of("t1", "v1", "t2", "v2");
+        property.setTags(tagMap);
+        PlainCommand nonEmptyCommand = new PropertyCommand(property);
+        CommandMethod.send(nonEmptyCommand);
+
+        Map<String, String> emptyTagMap = ImmutableMap.of("t1", "v1-new", "t2", emptyValue);
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Bad naming. It's not necessary to assign the value to a variable if the variable name gives no information additionally to setter name

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            dataProvider = "emptyValues"
+    )
+    public void emptyAndNonEmptyTagTest(String emptyValue) {
+        String propertyType = Mocks.propertyType();
+        String entityName = Mocks.entity();
+        Property property = new Property(propertyType, entityName);
+        Map<String, String> tagMap = ImmutableMap.of("t1", "v1", "t2", "v2");
+        property.setTags(tagMap);
+        PlainCommand nonEmptyCommand = new PropertyCommand(property);
+        CommandMethod.send(nonEmptyCommand);
```
Comment (raipc) [inline src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Response must be checked as well

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 11 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: image
There is a conflict without boxing

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 13 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Check added

File: src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java (hunk 12 above)
Comment (Arthasasyan) [reply src/test/java/com/axibase/tsd/api/method/property/command/EmptyTagTest.java]: Variables of Immutable map are deleted

**Comment by raipc** [comment]
Test is infinitely blocked if tag value contains line separator. This issue will be discussed.

**Pull request merged**